infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/ServerChecker.java

```diff
@@ -0,0 +1,134 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2017 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.workspace.infrastructure.docker.server;
+
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.slf4j.Logger;
+
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Checks availability of a server.
+ *
+ * @author Alexander Garagatyi
+ */
+public abstract class ServerChecker {
+    private static final Logger LOG = getLogger(ServerChecker.class);
+
+    private final String                    machineName;
+    private final String                    serverRef;
+    private final long                      period;
+    private final long                      deadLine;
+    private final CompletableFuture<String> reportFuture;
+
+    private Timer timer;
+
+    /**
+     * Creates server checker instance.
+     *
+     * @param machineName
+     *         name of machine to whom the server belongs
+     * @param serverRef
+     *         reference of the server
+     * @param period
+     *         period between unsuccessful availability checks, measured in {@code timeUnit}
+     * @param timeout
+     *         max time allowed for the server availability checks to last before
+     *         server is treated unavailable, measured in {@code timeUnit}
+     * @param timeUnit
+     *         measurement unit for {@code period} and {@code timeout} parameters
+     */
+    protected ServerChecker(String machineName,
+                            String serverRef,
+                            long period,
+                            long timeout,
+                            TimeUnit timeUnit) {
+        this.machineName = machineName;
+        this.serverRef = serverRef;
+        this.period = TimeUnit.MILLISECONDS.convert(period, timeUnit);
+        this.reportFuture = new CompletableFuture<>();
+        this.deadLine = System.currentTimeMillis() + TimeUnit.MILLISECONDS.convert(timeout, timeUnit);
+    }
+
+    /**
+     * Starts server availability checking, which will be stopped when server become available or
+     * checking times out.
+     */
+    public void start() {
+        if (timer != null) {
+            throw new IllegalStateException("ServerChecker is already started");
+        }
+        timer = new Timer("Server-" + serverRef +
```

Comment:
I'd say it perfectly fits the problem.
Moreover it allows you to call stop once!